### PR TITLE
Code Review Assignment 2 - Ruby 

### PR DESCRIPTION
**Pull request by andreeaiulia13**
This is a software for a car cleaning service.
It is a ruby program which determines the pickup date and time, knowing the functioning schedule of the Car Cleaning Service.

### CODE REVIEW

File: car_cleaning_service.rb [hunk 1]
```diff
@@ -0,0 +1,153 @@
+class Car
+  attr_reader :user_name
+  attr_reader :drop_time
+  attr_accessor :pickup_time
+
+  @@cars = Array.new
```
Comment (VictorMotogna) [inline car_cleaning_service.rb]: Having a list of cars inside the car model is a little confusing. Maybe having this variable outside the class:

```ruby
cars = []

class Car
  def initialize; end
end

cars << Car.new() # call this multiple times to add cars
```

File: car_cleaning_service.rb [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+  pres = 0
+  Car.all.each {|car| pres+=1 if ((car.drop_time > time && car.drop_time < time + 7200) || ( car.pickup_time > time && car.pickup_time < time + 7200))}
+  return true if pres == 2
+  return false
+end
+
+def add_appointment(time)
+  puts "What is your name?"
+  name = gets.chomp
+  if (name[/[a-zA-Z]*/] != name)
```
Comment (VictorMotogna) [inline car_cleaning_service.rb]: Maybe having a regex & using `match` would be cleaner here

File: car_cleaning_service.rb [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+  else 
+    add_appointment(Time.now)
+  end
+end
+
+def show_appointments 
+  Car.all.each {|car| puts "#{car.user_name}, #{car.drop_time.strftime("%d %B %Y %H:%M")}, #{car.pickup_time.strftime("%d %B %Y %H:%M")}" if car.pickup_time > Time.now}
+end
+
+def check_car_status
```
Comment (VictorMotogna) [inline car_cleaning_service.rb]: I think you should try to space lines a little bit, so it would be easier to read. Having spaces before ifs/loops, separating blocks of code, etc..

File: car_cleaning_service.rb [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+  pres = 0
+  puts "What is your name?"
+  name = gets.chomp
+  puts ""
+  Car.all.each{ |car|
```
Comment (VictorMotogna) [inline car_cleaning_service.rb]: you would already have access to list of cars here (@@cars), so you could use that. in Rails, if you'd already have this list and use `Car.all` that would make a new query to the database, which is unnecessary

File: car_cleaning_service.rb [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    if (car.user_name == name)
+      pres = 1
```
Comment (VictorMotogna) [inline car_cleaning_service.rb]: I think that `pres` is not really an intuitive name, and also you could change it to a boolean, it would be easier to understand it's purpose

File: car_cleaning_service.rb [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+  puts "Choose a month: \n1-January \n2-February\n3-March\n4-April\n5-May\n6-June\n7-July\n8-August\n9-September\n10-October\n11-November\n12-December"
+  month = gets.chomp.to_i 
+  puts "Which day?"
+  day = gets.chomp.to_i
+  puts "At what time?(hour)"
+  hour = gets.chomp.to_i
+  puts "minutes:"
+  min = gets.chomp.to_i
+  puts ""
+  if Time.new(2021, month, day, hour, min, 0) < Time.now
```
Comment (VictorMotogna) [inline car_cleaning_service.rb]: Maybe having a `case when` here wold make things more readable.

File: car_cleaning_service.rb [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    show_appointments
+  when "5"
+    puts "Thank you for choosing our car cleaning service." 
+    puts "We are waiting for you to come back, don't forget: A clean car is a happy car"
+    puts "Have a good one!"
+    break
+  else
+    puts "Please enter a number from 1-5 (1,2,3,4,5)"
+  end
+end
```
Comment (VictorMotogna) [inline car_cleaning_service.rb]: I think that a little refactoring for spacing would help: newlines between blocks/ifs/loops, adding a newline at the end of the file

**Review by VictorMotogna — CHANGES_REQUESTED** [review]

File: car_cleaning_service.rb [hunk 8]
```diff
@@ -0,0 +1,153 @@
+class Car
+  attr_reader :user_name
+  attr_reader :drop_time
```
Comment (vaneaclaudia) [inline car_cleaning_service.rb]: you can add `user_name` and `drop_time` on the same line like this:

`attr_reader :user_name, drop_time`

File: car_cleaning_service.rb (hunk 4 above)
Comment (vaneaclaudia) [inline car_cleaning_service.rb]: try do use `Car.all.each do ...end` when you have a multi line block. 
Usually for multi line blocks it is recommended to use do..end block and for single line block {}

File: car_cleaning_service.rb [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+  Car.all.each {|car| puts "#{car.user_name}, #{car.drop_time.strftime("%d %B %Y %H:%M")}, #{car.pickup_time.strftime("%d %B %Y %H:%M")}" if car.pickup_time > Time.now}
+end
+
+def check_car_status
+  pres = 0
+  puts "What is your name?"
+  name = gets.chomp
+  puts ""
+  Car.all.each{ |car|
+    if (car.user_name == name)
```
Comment (vaneaclaudia) [inline car_cleaning_service.rb]: instead of having so many `if` conditions you can use here `next unless car.user_name == name`
also no need to use parentheses

File: car_cleaning_service.rb (hunk 7 above)
Comment (vaneaclaudia) [reply car_cleaning_service.rb]: I agree with @VictorMotogna. Empty lines between block will improve readability

**Review by vaneaclaudia — CHANGES_REQUESTED** [review]

File: Controler/controler.rb [hunk 10]
```diff
@@ -0,0 +1,116 @@
+require "/home/andreea/Desktop/Assignment2_CarCleaningService/Model/model.rb"
```
Comment (VictorMotogna) [inline Controler/controler.rb]: You should try to use `require_relative`. If I pull this code and run it on my machine, it will not work, because I don't have this path on my machine

File: Model/model.rb [hunk 11]
```diff
@@ -0,0 +1,30 @@
+$cars = []
```
Comment (VictorMotogna) [inline Model/model.rb]: You don't need to add `$` to variable names

File: Controler/controler.rb [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+end
+
+def set_time(user, time)
+  puts "Dear, #{user} your car will be ready on #{time.strftime("%d %B %Y")} at #{time.strftime("%H:%M")}."
+end
+
+def check_schedule(time) 
+  count_cars_in_a_time_interval = 0
+  
+  $cars.each {|car| count_cars_in_a_time_interval += 1 if ((car.drop_time > time && car.drop_time < time + 7200) || (car.pickup_time > time && car.pickup_time < time + 7200))}
```
Comment (VictorMotogna) [inline Controler/controler.rb]: you could add a space before & after `{}`:
```ruby
cars.each { |car| c.do_something }
```

File: Controler/controler.rb [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+  return false
+end
+
+def add_appointment(time)
+  regular_name = /([\w\-\']{2,})([\s]+)([\w\-\']{2,})/
+
+  puts "What's your name?"
+  name = gets.chomp
+
+  if !name.match(regular_name)
```
Comment (VictorMotogna) [inline Controler/controler.rb]: you can refactor to use `unless name.match`

File: Controler/controler.rb [hunk 14]
```diff
@@ -0,0 +1,116 @@
+require "/home/andreea/Desktop/Assignment2_CarCleaningService/Model/model.rb"
+
```
Comment (vaneaclaudia) [inline Controler/controler.rb]: if you want to follow the MVC pattern, then you should have your folders named like controllers/models/views

File: Controler/controler.rb [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+  when check_opening_time(Time.new(2021, month, day, hour, min, 0))
+    show_schedule
+		
+  when check_schedule(Time.new(2021, month, day, hour, min, 0))
+    puts "There is no available car wash machine at this moment. Please check our appointments schedule and try another date or time"
+	
+  else 
+    add_appointment(Time.new(2021, month, day, hour, min, 0))
+  
+	end
```
Comment (VictorMotogna) [inline Controler/controler.rb]: the indentation of this `end` seems a little bit off

File: View/view.rb [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+require "/home/andreea/Desktop/Assignment2_CarCleaningService/Model/model.rb"
+
+puts ""
+puts "        WELCOME TO OUR CAR CLEANING SERVICE! \n"
+puts "       Caring for your vehicle is our business"
+puts "              Clean cars run better."
+
+option = 0
+
+while(option != 5)
```
Comment (VictorMotogna) [inline View/view.rb]: it seems like the `6` is the exit command, not 5

File: Controler/controler.rb [hunk 17]
```diff
@@ -0,0 +1,116 @@
+require "/home/andreea/Desktop/Assignment2_CarCleaningService/Model/model.rb"
+
+def check_opening_time(time)
+ return ((time.sunday?) || (time.hour < 8) || ((time + 7200).hour >= 18) || ((time + 7200).hour == 0)|| ((time.saturday?) && ((time + 7200).hour >= 16)))
```
Comment (vaneaclaudia) [inline Controler/controler.rb]: You don't need to use return since it's the only line of code you have. You should use `return` keyword when you have multiple blocks of code and you want to return from the method if some conditions are met. 
Also, this line with conditions is very hard to follow. I suggest splitting these in 2 methods

File: Controler/controler.rb [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+  puts "Dear, #{user} your car will be ready on #{time.strftime("%d %B %Y")} at #{time.strftime("%H:%M")}."
+end
+
+def check_schedule(time) 
+  count_cars_in_a_time_interval = 0
+  
+  $cars.each {|car| count_cars_in_a_time_interval += 1 if ((car.drop_time > time && car.drop_time < time + 7200) || (car.pickup_time > time && car.pickup_time < time + 7200))}
+  
+  return true if count_cars_in_a_time_interval >= 2
+  return false
```
Comment (vaneaclaudia) [inline Controler/controler.rb]: same remark here regarding redundant `return`

File: spec/car_service_spec.rb [hunk 19]
```diff
@@ -0,0 +1,91 @@
+require "/home/andreea/Desktop/Assignment2_CarCleaningService/Controler/controler.rb"
+
+describe "check open times" do
+
+	context "when sunday" do
```
Comment (VictorMotogna) [inline spec/car_service_spec.rb]: the indentation in this file seems to be 4 spaces, it looks like it's different than the rest of the files

**Review by VictorMotogna — CHANGES_REQUESTED** [review]

File: Controler/controler.rb [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+  elsif check_schedule(Time.now)
+    puts "There is no available car wash machine at this moment. Would you like to make an appointment?"
+  
+  else 
+    add_appointment(Time.now)
+  end
+end
+
+def show_appointments 
+  $cars.each {|car| puts "Drop time: #{car.drop_time.strftime("%d %B %Y %H:%M")}, Pick up time: #{car.pickup_time.strftime("%H:%M")}" if car.pickup_time > Time.now}
```
Comment (vaneaclaudia) [inline Controler/controler.rb]: I suggest you use a `do..end` block here, it will be easier to follow the code since the line is so long

File: spec/car_service_spec.rb [hunk 21]
```diff
@@ -0,0 +1,91 @@
+require "/home/andreea/Desktop/Assignment2_CarCleaningService/Controler/controler.rb"
+
+describe "check open times" do
```
Comment (vaneaclaudia) [inline spec/car_service_spec.rb]: please use single quotes instead of double quotes, usually double quotes are used when you have interpolation

File: spec/car_service_spec.rb [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+
+	context "when sunday" do
+		time = Time.new(2021, 04, 25, 9, 00, 0)
```
Comment (vaneaclaudia) [inline spec/car_service_spec.rb]: this could be a let variable

let(:time) { Time.new(2021, 04, 25, 9, 00, 0) }

File: spec/car_service_spec.rb [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+		
+		it "return true" do
```
Comment (vaneaclaudia) [inline spec/car_service_spec.rb]: same remark using single quotes

File: spec/car_service_spec.rb [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+			expect(check_opening_time(time)).to eq (true)
+		end
+	end
+
+	context "when saturday after 16:00" do
+		time = Time.new(2021, 04, 24, 16, 00, 0)
```
Comment (vaneaclaudia) [inline spec/car_service_spec.rb]: same remark using let variable

File: spec/car_service_spec.rb [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+		end
+	end
+end
+
+describe "check cars wash machines availability" do 
+
+	context "when both car wash machines are available" do
+		before(:each) do 
+			$cars = $cars.clear()
+		end
```
Comment (vaneaclaudia) [inline spec/car_service_spec.rb]: since you have only one line of code you can do something like:

`before { $cars = $cars.clear() }`

**Review by vaneaclaudia — CHANGES_REQUESTED** [review]

File: Model/model.rb (hunk 11 above)
Comment (andreeaiulia13) [reply Model/model.rb]: It gives me an error if I don't use "$" :undefined local variable or method `cars' for main:Object (NameError)

Comment (VictorMotogna) [reply Model/model.rb]: ok, I see, my bad. in that case, you can use it like this

File: controllers/controller.rb [hunk 26]
```diff
@@ -0,0 +1,133 @@
...
+
+  puts 'At what time?(hour)'
+  hour = gets.chomp.to_i
+
+  puts 'Minutes:'
+  min = gets.chomp.to_i
+  puts ''
+	
+  case 
+  when Time.new(2021, month, day, hour, min, 0) < Time.now
```
Comment (VictorMotogna) [inline controllers/controller.rb]: you can move `Time.new(2021, month, day, hour, min, 0)` to a variable, so you won't instantiate for each check

File: spec/service_spec.rb [hunk 27]
```diff
@@ -0,0 +1,86 @@
+require_relative '../controllers/controller.rb'
+
+describe 'check open times' do
+
```
Comment (VictorMotogna) [inline spec/service_spec.rb]: the indentation in this file seems to be 4 spaces

File: spec/service_spec.rb [hunk 28]
```diff
@@ -0,0 +1,86 @@
...
+			$cars = $cars.clear()
+			$cars << Car.new("Mary Test", Time.now, "AN12345")
+			$cars << Car.new("Johnny Test", Time.now, "DN45678")
+		end
+		
+		it 'return true' do 
+			expect(check_schedule(Time.now)).to eq (true)
+		end
+	end
+end
```
Comment (VictorMotogna) [inline spec/service_spec.rb]: you'll need to add a newline at the end of the file

**Review by VictorMotogna — CHANGES_REQUESTED** [review]

**Review by VictorMotogna — APPROVED** [review]

**Review by vaneaclaudia — APPROVED** [review]
Good job! 👍 💯
